From b3f70bb672516e431708095deda1a24fd250401f Mon Sep 17 00:00:00 2001
From: Kiran Rudramuni <kchitrik@codeaurora.org>
Date: Fri, 20 Mar 2015 16:23:39 -0700
Subject: [PATCH] msm_shared: hdmi: provide support to check for cable status

For some uses cases, hdmi cable status is needed to be known at boot up
to take further actions based of cable status. Provide an API which can
give the current status of cable connection.

Change-Id: I889cb7a41b8d4888d8c6aac3dec8da6215b3e2c6
---
 platform/msm_shared/mdss_hdmi.c | 98 ++++++++++++++++++++++++++++++++++++-----
 1 file changed, 86 insertions(+), 12 deletions(-)

diff --git a/platform/msm_shared/mdss_hdmi.c b/platform/msm_shared/mdss_hdmi.c
index 91997a6..af40586 100644
--- a/platform/msm_shared/mdss_hdmi.c
+++ b/platform/msm_shared/mdss_hdmi.c
@@ -35,6 +35,9 @@
 #include <platform/iomap.h>
 
 static struct msm_fb_panel_data panel;
+static bool hdmi_power_enabled;
+static bool hdmi_panel_clock_enabled;
+static bool hdmi_pll_clock_enabled;
 
 /* Supported HDMI Audio channels */
 #define MSM_HDMI_AUDIO_CHANNEL_2		1
@@ -567,35 +570,58 @@ static void mdss_hdmi_audio_playback(void)
 
 static int mdss_hdmi_panel_clock(uint8_t enable)
 {
-	return target_hdmi_panel_clock(enable);
+	int ret = NO_ERROR;
+
+	if (hdmi_panel_clock_enabled)
+		return ret;
+
+	ret = target_hdmi_panel_clock(enable);
+
+	hdmi_panel_clock_enabled = true;
+
+	return ret;
 }
 
 static int mdss_hdmi_pll_clock(uint8_t enable, struct msm_panel_info *pinfo)
 {
-	return target_hdmi_pll_clock(enable, pinfo);
+	int ret = NO_ERROR;
+
+	if (hdmi_pll_clock_enabled)
+		return ret;
+
+	ret = target_hdmi_pll_clock(enable, pinfo);
+
+	hdmi_pll_clock_enabled = true;
+
+	return ret;
 }
 
 static int mdss_hdmi_enable_power(uint8_t enable)
 {
-        int ret = NO_ERROR;
+	int ret = NO_ERROR;
+
+	if (hdmi_power_enabled)
+		return ret;
 
-        ret = target_ldo_ctrl(enable);
-        if (ret) {
+	ret = target_ldo_ctrl(enable);
+	if (ret) {
 		dprintf(CRITICAL, "LDO control enable failed\n");
 		goto bail_ldo_fail;
-        }
+	}
 
-        ret = target_hdmi_regulator_ctrl(enable);
-        if (ret) {
+	ret = target_hdmi_regulator_ctrl(enable);
+	if (ret) {
 		dprintf(CRITICAL, "hdmi regulator control enable failed\n");
 		goto bail_regulator_fail;
-        }
+	}
 
 	ret = target_hdmi_gpio_ctrl(enable);
 	if (ret) {
 		dprintf(CRITICAL, "hdmi gpio control enable failed\n");
 		goto bail_gpio_fail;
-        }
+	}
+
+	hdmi_power_enabled = true;
 
 	dprintf(SPEW, "HDMI Panel power %s done\n", enable ? "on" : "off");
 
@@ -821,11 +847,52 @@ static void mdss_hdmi_panel_init(struct msm_panel_info *pinfo)
 	pinfo->lcdc.dual_pipe  = 0;
 }
 
+static int mdss_hdmi_cable_status(void)
+{
+	uint32_t reg_val, cable_status;
+
+	mdss_hdmi_set_mode(true);
+
+	/* Enable USEC REF timer */
+	writel(0x0001001B, HDMI_USEC_REFTIMER);
+
+	/* set timeout to 4.1ms (max) for hardware debounce */
+	reg_val = readl(HDMI_HPD_CTRL) | 0x1FFF;
+
+	/* enable HPD circuit */
+	writel(reg_val | BIT(28), HDMI_HPD_CTRL);
+
+	writel(BIT(2) | BIT(1), HDMI_HPD_INT_CTRL);
+
+	/* Toggle HPD circuit to trigger HPD sense */
+	reg_val = readl(HDMI_HPD_CTRL) | 0x1FFF;
+	writel(reg_val & ~BIT(28), HDMI_HPD_CTRL);
+	writel(reg_val & BIT(28), HDMI_HPD_CTRL);
+
+	mdelay(20);
+
+	cable_status = (readl(HDMI_HPD_INT_STATUS) & BIT(1)) >> 1;
+
+	dprintf(INFO, "hdmi cable %s\n",
+		cable_status ? "connected" : "not connected");
+
+	writel(BIT(0), HDMI_HPD_INT_CTRL);
+	writel(reg_val & ~BIT(28), HDMI_HPD_CTRL);
+
+	mdss_hdmi_set_mode(false);
+
+	return cable_status;
+}
+
 static int mdss_hdmi_update_panel_info(void)
 {
 	mdss_hdmi_set_mode(true);
-	mdss_hdmi_read_edid();
-	mdss_hdmi_parse_res();
+
+	if (!mdss_hdmi_read_edid())
+		mdss_hdmi_parse_res();
+	else
+		mdss_hdmi_video_fmt = DEFAULT_RESOLUTION;
+
 	mdss_hdmi_set_mode(false);
 
 	mdss_hdmi_panel_init(&(panel.panel_info));
@@ -851,6 +918,13 @@ void mdss_hdmi_display_init(uint32_t rev, void *base)
 
 	msm_display_init(&panel);
 }
+int target_hdmi_check_cable_status(void)
+{
+	mdss_hdmi_enable_power(1);
+	mdss_hdmi_panel_clock(1);
+
+	return mdss_hdmi_cable_status();
+}
 
 static int mdss_hdmi_video_setup(void)
 {
-- 
1.8.2.1

